include/seqan3/core/bit_manipulation.hpp

```diff
@@ -77,33 +81,94 @@ constexpr size_t next_power_of_two(size_t n)
     return n + 1;
 }
 
-/*!\brief Returns the position of the most significant bit (counting from right to left).
+/*!\brief Returns the number of 1-bits.
+* \ingroup core
+*
+* \param[in] n An unsigned integer.
+*
+* \returns The number of 1-bits.
+*/
+template <std::UnsignedIntegral unsigned_t>
+constexpr uint8_t popcount(unsigned_t n)
+{
+#if defined(__GNUC__)
+    if constexpr (sizeof(unsigned_t) == sizeof(unsigned long long))
+        return __builtin_popcountll(n);
+    else if constexpr (sizeof(unsigned_t) == sizeof(unsigned long))
+        return __builtin_popcountl(n);
+    else
+        return __builtin_popcount(n);
+#else
+    return sdsl::bits::cnt(n);
+#endif
+}
+
+/*!\brief Returns the number of leading 0-bits, starting at the most significant bit position.
+* \ingroup core
+*
+* \param[in] n An unsigned integer.
+*
+* \attention *n = 0* is a special case and is undefined behaviour.
+*
+* \returns The number of leading 0-bits.
+*/
+template <std::UnsignedIntegral unsigned_t>
+constexpr uint8_t count_leading_zeros(unsigned_t n)
+{
+    assert(n > 0); // n == 0 might have undefined behaviour
+#if defined(__GNUC__)
+    if constexpr (sizeof(unsigned_t) == sizeof(unsigned long long))
+        return __builtin_clzll(n);
+    else if constexpr (sizeof(unsigned_t) == sizeof(unsigned long))
+        return __builtin_clzl(n);
+    else
+        return __builtin_clz(n) + sizeof_bits<unsigned_t> - sizeof_bits<unsigned int>;
+#else
+    return sizeof_bits<unsigned_t> - sdsl::bits::hi(n) - 1;
+#endif
+}
+
+/*!\brief Returns the number of trailing 0-bits, starting at the least significant bit position.
  * \ingroup core
  *
  * \param[in] n An unsigned integer.
  *
  * \attention *n = 0* is a special case and is undefined behaviour.
  *
- * \returns The position of the most significant bit.
+ * \returns The number of trailing 0-bits.
  */
 template <std::UnsignedIntegral unsigned_t>
-constexpr uint8_t bit_scan_reverse(unsigned_t n)
+constexpr uint8_t count_trailing_zeros(unsigned_t n)
 {
     assert(n > 0); // n == 0 might have undefined behaviour
 #if defined(__GNUC__)
     if constexpr (sizeof(unsigned_t) == sizeof(unsigned long long))
-        return sizeof_bits<unsigned long long> - __builtin_clzll(n) - 1;
+        return __builtin_ctzll(n);
     else if constexpr (sizeof(unsigned_t) == sizeof(unsigned long))
-        return sizeof_bits<unsigned long> - __builtin_clzl(n) - 1;
+        return __builtin_ctzl(n);
     else
-        return sizeof_bits<unsigned> - __builtin_clz(n) - 1;
+        return __builtin_ctz(n);
 #else
-    uint8_t i = 0;
-    for (; n != 0; n >>= 1, ++i);
-    return i - 1;
+    return sdsl::bits::lo(n);
 #endif
 }
 
+/*!\brief Returns the position of the most significant bit (counting from right to left).
```

Comment:
> 0 based position? (add this information)

Ok.
> 
> also: does this position depend on little /big endianness?

Builtins are defined as `starting at the most/least significant bit position.` So yes, it depends since you interpret the bits differently.
> 
> and: do we need bit_scan_reverse at all?
> isn't `sizeof_bits<unsigned_t> - count_leading_zeros(n) - 1` == `count_trailing_zeros(n)`?

```cpp
uint16_t x = 0b0000'0001'0000'0000
uint8_t pos = sizeof_bits<uint16_t> - count_leading_zeros(x) - 1; 
//    8 = 16 - 7 - 1
uint8_t ctz = count_trailing_zeros(x);
//    ctz = 8
```
But
```cpp
uint16_t x = 0b0000'0001'0000'0010
uint8_t pos = sizeof_bits<uint16_t> - count_leading_zeros(x) - 1; 
//    8 = 16 - 7 - 1
uint8_t ctz = count_trailing_zeros(x);
//    ctz = 1
```